### ISSUE TITLE
checkArity confusingly raises line of outer caller for nested calls

### ISSUE THREAD

**Issue by andrewvc** [jruby/jruby#5103]
### Environment

Provide at least:
* JRuby version: 9.1.13.0
* Mac OSX

### Expected Behavior

I have code something like:

```
#Java interface
interface MyJavaInterface {
  void interfaceMethod(argItTakes)
}

#Ruby
class MyClass
  include org.example.MyJavaInterface

  # Note that we accidentally specify an extra arg here
  def interfaceMethod(argItTakes, extraArgThatIsAccident)
  end

  def someOtherMethod
    callJavaMethodThatInvokesAboveInterfaceMethodButWithOnlyOneArg
  end
end
```

### Actual Behavior

JRuby generates an `ArgumentError: wrong number of arguments (1 for 2)` that points to the line in `someOtherMethod`, when the real error is that we implemented the interface incorrectly. The stack trace shouldn't filter out the java in this case, but do the full stack trace to expose this.

**Comment by kares** [comment]
basically boils down to having some-form of static Java checks in a dynamic language (Ruby).
at the time you include the interface its methods aren't implemented (and might actually never be - that is fine in a dynamic language). when the proxy is generated there might be some room for args checking and maybe issuing a warning, that might be useful to some extend - also slightly tricky to get right since in some cases dispatch reaches `method_missing` and you might not have all of the methods implemented and if you do there's no way knowing which will be actually used (called) so we might warn too much ...

as for the stack-trace could you please clarify what you mean. 
doubt JRuby wants to switch trace type based on what calls triggered the issue.

**Comment by andrewvc** [comment]
@kares yes, this does seem hard to get right. The thing is, it is a bad user experience at the moment because JRuby quite confidently tells you you have an args issue at a given line, when the problem is really somewhere else.

I don't know if there is an efficient easy way to solve this. It would be nice if possible however.

The current stack, roughly speaking, is RrubyMethod(A) -> JavaMethod(B) -> RubyMethodThatIncorrectlyImplementsJavaInterface(C).

The error message today makes it seem like A is where the bug is, when the real problem is  somewhere between B and C

**Comment by headius** [comment]
I feel like the real problem here is that we have no context for the failed call, because the dispatch logic for the Ruby-based interface impl doesn't insert any context. It's a sort of hidden method_missing dispatch happening under the covers, but since it's implemented in "native" code it doesn't show up in the Ruby trace.

The dispatch logic here could perhaps be improved to provide a better error, perhaps something like "interface method SomeIfc.someMethod was not implemented in class SomeRubyClass".

**Comment by headius** [comment]
A similar issue is documented in #6335. The case here is slightly different, since we are simply not defining the expected interface method on the generated class, but the basic cause (arity check happens during call plumbing and not inside the method) is the same.
